concordium-consensus/src/Concordium/GlobalState/Basic/TreeState.hs

```diff
@@ -399,19 +410,34 @@ instance (bs ~ BlockState m, BS.BlockStateStorage m, Monad m, MonadIO m, MonadSt
         transactionTable .= newTT
         pendingTransactions .= newPT
 
-    wipeNonFinalizedTransactions = do
-        let consNonFin (_, Finalized{}) = id
-            consNonFin (bi, _) = (bi :)
-            isFin (_, Finalized{}) = True
-            isFin _ = False
-        -- The motivation for using foldr here is that the list will be consumed by iteration
-        -- almost immediately, so it is reasonable to build it lazily.
-        oldTransactions <- HM.foldr consNonFin [] <$> use (transactionTable . ttHashMap)
-        -- Filter to only the finalized transactions.
-        transactionTable %= (ttHashMap %~ HM.filter isFin)
-                . (ttNonFinalizedTransactions %~ fmap (anftMap .~ Map.empty))
-                . (ttNonFinalizedChainUpdates %~ fmap (nfcuMap .~ Map.empty))
-        return oldTransactions
+    clearOnProtocolUpdate = do
+        -- clear pending blocks
+        possiblyPendingTable .= HM.empty
+        possiblyPendingQueue .= MPQ.empty
+        -- clear non-finalized blocks.
+        branches .=! Seq.empty
+        let nonFinDead TS.BlockPending{} = TS.BlockDead
+            nonFinDead TS.BlockAlive{} = TS.BlockDead
+            nonFinDead o = o
+        blockTable %=! fmap nonFinDead
+        -- mark transactions that are not finalized as received since
+        -- they are no longer committed to any blocks.
+        transactionTable
+            . ttHashMap
+            %=! HM.map
+                ( \(bi, s) -> case s of
+                    Committed{..} -> (bi, Received{..})
+                    _ -> (bi, s)
+                )
+
+    clearAfterProtocolUpdate = do
+        transactionTable .=! emptyTransactionTable
+        pendingTransactions .=! emptyPendingTransactionTable
+        nextGenesisInitialState .=! Nothing
+        BS.collapseCaches
```

Comment:
Again, this is the basic state, which really does not matter so I spent very little time thinking about it. Especially since we are removing it in a couple of weeks.

In any case, I added so that finalized transactions are retained here.